cpp/src/arrow/filesystem/s3_test_util.cc

```diff
@@ -0,0 +1,174 @@
+std::string MinioTestServer::connect_string() const { return impl_->connect_string_; }
+
+std::string MinioTestServer::access_key() const { return impl_->access_key_; }
+
+std::string MinioTestServer::secret_key() const { return impl_->secret_key_; }
+
+Status MinioTestServer::Start() {
+  const char* connect_str = std::getenv(kEnvConnectString);
+  const char* access_key = std::getenv(kEnvAccessKey);
+  const char* secret_key = std::getenv(kEnvSecretKey);
+  if (connect_str && access_key && secret_key) {
+    // Use external instance
+    impl_->connect_string_ = connect_str;
+    impl_->access_key_ = access_key;
+    impl_->secret_key_ = secret_key;
+    return Status::OK();
+  }
+
+  ARROW_ASSIGN_OR_RAISE(impl_->temp_dir_, TemporaryDir::Make("s3fs-test-"));
+
+  // Get a copy of the current environment.
+  // (NOTE: using "auto" would return a native_environment that mutates
+  //  the current environment)
+  bp::environment env = boost::this_process::environment();
+  env["MINIO_ACCESS_KEY"] = kMinioAccessKey;
+  env["MINIO_SECRET_KEY"] = kMinioSecretKey;
+
+  impl_->connect_string_ = GenerateConnectString();
+
+  auto exe_path = bp::search_path(kMinioExecutableName);
+  if (exe_path.empty()) {
+    return Status::IOError("Failed to find minio executable ('", kMinioExecutableName,
+                           "') in PATH");
+  }
+
+  try {
+    // NOTE: --quiet makes startup faster by suppressing remote version check
+    impl_->server_process_ = std::make_shared<bp::child>(
+        env, exe_path, "server", "--quiet", "--compat", "--address",
+        impl_->connect_string_, impl_->temp_dir_->path().ToString());
+  } catch (const std::exception& e) {
+    return Status::IOError("Failed to launch Minio server: ", e.what());
+  }
+  return Status::OK();
+}
+
+Status MinioTestServer::Stop() {
+  if (impl_->server_process_ && impl_->server_process_->valid()) {
+    // Brutal shutdown
+    impl_->server_process_->terminate();
+    impl_->server_process_->wait();
+  }
+  return Status::OK();
+}
+
+struct MinioTestEnvironment::Impl {
+  std::function<Future<std::shared_ptr<MinioTestServer>>()> server_generator_;
+
+  Result<std::shared_ptr<MinioTestServer>> LaunchOneServer() {
+    auto server = std::make_shared<MinioTestServer>();
+    RETURN_NOT_OK(server->Start());
+    return server;
+  }
+};
+
+MinioTestEnvironment::MinioTestEnvironment() : impl_(new Impl) {}
+
+MinioTestEnvironment::~MinioTestEnvironment() = default;
+
+void MinioTestEnvironment::SetUp() {
+  auto pool = ::arrow::internal::GetCpuThreadPool();
+
+  auto launch_one_server = []() -> Result<std::shared_ptr<MinioTestServer>> {
+    auto server = std::make_shared<MinioTestServer>();
+    RETURN_NOT_OK(server->Start());
+    return server;
+  };
+  impl_->server_generator_ = [pool, launch_one_server]() {
+    return DeferNotOk(pool->Submit(launch_one_server));
+  };
+  impl_->server_generator_ =
+      MakeReadaheadGenerator(std::move(impl_->server_generator_), pool->GetCapacity());
```

Comment:
Would a fixed readahead level be better? On my system for instance I think this would instantly spawn 18 minio instances…Though maybe that's fine since we need an instance per test.

Reply:
On my system it spawns 24 of them :-)

Reply:
So, yes, it's fine IMHO. They'll just sit waiting for requests.